### ISSUE TITLE
implement sparse matrix factorization in mxnet

### ISSUE THREAD

**Issue by xlvector** [dmlc/mxnet#2408]
Consider a common problem in recommender system:

r_ui = dot(p_u, q_i)

In mxnet, we can do this by:

pu = Embedding(u)
qi = Embedding(i)
qi = Transpose(qi)
rui = dot(pu, pi)
loss = LinearRegessionOutput(rui)

However, dot operator in mxnet is a matrix multiplication. Thus, if shape of u is (batch_size, 1), shape of i is (batch_size, 1), then, shape of pu is (batch_size, num_hidden), shape of qi is (batch_size, num_hidden). at last, the shape of rui is (batch_size, batch_size).  However, in above problem, we want a dot operator which can generate rui with shape (batch_size, 1).

So, is there easy way to solve this problem without adding a new operator?

**Comment by sxjscience** [comment]
Use elementwise product first and then use sum over axis?

**Comment by xlvector** [comment]
I does not find elementwise product op, I only find elementwise sum op.

**Comment by Godricly** [comment]
maybe concated them first and do a matrix multiplication with (1,1)?

**Comment by xlvector** [comment]
@Godricly I will have a try.

**Comment by xlvector** [comment]
(batch_size * n_hidden, 1)^T * (batch_size * n_hidden, 1) = (1, 1)

But I need (batch_size, 1)...

Another hack way is always set batch_size = 1, this is not good.

**Comment by sxjscience** [comment]
@xlvector Directly multiply will work. E.g c=a*b. Here * is elementwise

**Comment by xlvector** [comment]
@sxjscience In http://mxnet.readthedocs.io/en/latest/packages/python/symbol.html, 
I only find broadcast_mul?

**Comment by sxjscience** [comment]
It's called something like symbol._mul.

**Comment by xlvector** [comment]
Finally make it work:  https://github.com/xlvector/mxnet/tree/svd/example/svd

Since its very easy, will not submit PR.
